### DISCUSSION TITLE
IconColumn won't work if it doesn't match a model field.

### DISCUSSION THREAD

**Discussion by maxacarvalho** [filamentphp/filament#7651] [Help]
Hi,

I'm trying to add a column that does not reflect a Model field (property).
It worked on the v2 but stopped with the v3.
This is how it looked on v2:
```php
...
IconColumn::make('buyer')
    ->label(Str::formatTitle(__('user.is_buyer')))
    ->options([
        'heroicon-o-check-circle' => fn ($state, Model|User $record): bool => $record->isBuyer(),
]),
...
```

And here is the v3 equivalent:
```php
IconColumn::make('buyer')
    ->label(Str::formatTitle(__('user.is_buyer')))
    ->icon(function (string $state, Model|User $record): string {
        if ($record->isBuyer()) {
            return 'heroicon-o-check-circle';
        }

        return '';
    }),
```

Note that the `buyer` does not match any table column (model property).  
If I change it to some of the existing model property names, it works.

**Comment by danharrin** [comment]
Feel free to open an issue with a reproduction repo for this, otherwise it probably wont be fixed

**Reply by maxacarvalho** [reply]
Hi @danharrin 
I created a repo with a fresh Laravel install and Filament, then started to add content for a better understanding of the issue.  
I didn't dig deep into the Filament Tables implementation but the issue I mentioned above might actually not be a bug.

As far as I understood, the Filament Tables attempts to get the data out of the column's `make()` argument. it uses the given parameter value to fetch the state from the model data that's loaded.
So, when an non-existent column name is given the logic cannot find anything to load.

While testing it I could see that the `->icon(fn (bool $state)...` callback is not even called in when the column's make argument doesn't match an existing Model property. I'd assume it happens because the Table builder ignores it.
I find odd that it wasn't the case with version 2 but it makes sense that it behaves like that.

After realising that I found a solution:

```php
IconColumn::make('seller')
    ->label(Str::formatTitle(__('user.is_seller')))
    ->alignCenter()
    ->color('success')
    ->getStateUsing(fn (Model|User $record): bool => $record->isSeller())
    ->icon(fn (bool $state): string => $state ? 'heroicon-o-check-circle' : ''),
```

Now the `icon()` callback is executed since I'm giving it a proper state to use.

Does that make sense, would you say it's like that by design or does it look like a bug?

**Reply by danharrin** [reply]
That does make sense, I would continue to use `getStateUsing()` if I were you

**Comment by adampatterson** [comment]
I also had this issue, and it worked v2 but not v3.

Oddly `tooltip()` was working fine.

I used `getStateUsing(fn(Model $record): Model => $record)`  

Was there a reason for this change? I didn't notice it in the update steps and `getStateUsing` isn't documented.
